notebooks/WFC3/point_spread_function/psf_utilities.py

```diff
@@ -0,0 +1,607 @@
+
+    # Download the PSF file if it doesn't exist.
+    if not os.path.exists(psf_path):
+        print('Downloading:', psf_url+psf_name)
+        urllib.request.urlretrieve(psf_url+psf_name, psf_path)
+
+    # Confirm that the file can be opened successfully.
+    try:
+        hdul = fits.open(psf_path, ignore_missing_end=True)
+        hdul.close()
+        print('Validation complete, the PSF file is readable.')
+    except IOError:
+        raise IOError('ERROR: Unable to open', psf_name)
+    except Exception as e:
+        raise Exception(e)
+
+    return psf_name
+
+
+def make_cutouts(image, star_ids, xis, yis, rpix, 
+                 scale_stars=True, sub_pixel=True, show_figs=True, verbose=True):
+
+    """
+    A function that extracts postage stamps or cutouts of sources from a 
+    science image provided an image, list of source IDs, x and y detector 
+    coordinates, and several optional parameters. The scale_stars parameter 
+    normalizes the peak flux of each source to unity, which is required 
+    when taking the mean or median in subsequent steps. The subpixel option 
+    aligns the cutouts at the subpixel level, rather than using integer 
+    array values, utilizing user provided x and y coordinates, typically 
+    from hst1pass or photutils. The show_figs and verbose options allow 
+    users to toggle diagnostic figures and messages on or off depending 
+    on the number of sources. As a check of the centroiding, the code 
+    will calculate the center of mass of the inner 5 pixels, which is 
+    similar to the initial guesses used in hst1pass. This provides a 
+    centroid for comparison with the user provided values. In general, 
+    the values should be very close to rpix but not equal exactly. The 
+    image is then shifted by the subpixel offset for the best alignment.
+
+    Parameters
+    ----------
+    image : np.ndarray
+        The np.ndarray containing the array of science data.
+    star_ids : list
+        A list of the integer stellar IDs from a catalog or arange.
+    xis : list
+        A list of the x-centroids of each star, assumed to be exact.
+    yis : list
+        A list of the y-centroids of each star, assumed to be exact.
+    rpix : int
+        An integer number of pixels for the cutout length and width.
+    scale_stars : bool
+        If True then the value of each pixel is divided by the maximum. 
+        This should generally be set to True to avoid improper weighting.
+    sub_pixel : bool
+        If True then a sub-pixel recentering is completed, using the
+        values provided by the user in the xis and yis lists, so that 
+        the stars centroid is located at exactly (rpix, rpix). This 
+        should generally be set to True.
+    show_figs : bool
+        If True then each cutout will be shown in the notebook.
+    verbose : bool
+        If True then additional diagnostic information is printed.
+
+    Returns
+    -------
+    image_array : np.ndarray
```

Comment:
Interestingly, the stack_cutouts function works with either a list or array input. I think in this case because it's providing a list of arrays to take the median of in stack_cutouts. I've added the suggested line of code in make_cutouts.